File: app/src/main/java/com/github/mattiadellepiane/gnssraw/listeners/FileLogger.java
package com.github.mattiadellepiane.gnssraw.listeners;

import android.icu.text.SimpleDateFormat;
import android.os.Environment;

import com.github.mattiadellepiane.gnssraw.R;
import com.github.mattiadellepiane.gnssraw.data.SharedData;

import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Date;

public class FileLogger extends MeasurementListener{

    private PrintWriter out;
    private File file;
    private final String FOLDER;
    private String currentFileName;

    public FileLogger(){
        FOLDER = SharedData.getInstance().getContext().getString(R.string.app_documents_folder);
    }

    @Override
    protected void initResources() {
        SimpleDateFormat formatter = new SimpleDateFormat("yyyy_MM_dd_HH_mm_ss");
        String now = formatter.format(new Date());
        String fileName = String.format(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOCUMENTS) + "%s_%s.txt", FOLDER + "/GNSS_Log", now);
        currentFileName = "GNSS_Log_" + now + ".txt";
        file = createFile(fileName);
        try {
            out = new PrintWriter(file);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Override
    protected void releaseResources() {
        if(out != null)
            out.close();
        if(SharedData.getInstance().getFilesFragment() != null)
            SharedData.getInstance().getFilesFragment().addFileView(currentFileName);
    }

    @Override
    protected void write(String s) {
        if(out != null)
            out.println(s);
    }

    private File createFile(String s){
        File f = new File(s);
        if(!f.getParentFile().exists()){
            f.getParentFile().mkdirs();
        }
        try {
            f.createNewFile();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return f;
    }
}

File: app/src/main/java/com/github/mattiadellepiane/gnssraw/listeners/FileLogger.kt
package com.github.mattiadellepiane.gnssraw.listeners

import com.github.mattiadellepiane.gnssraw.R
import com.github.mattiadellepiane.gnssraw.data.SharedData
import java.io.File
import android.os.Environment
import java.io.IOException
import java.io.PrintWriter
import android.icu.text.SimpleDateFormat
import java.util.*

class FileLogger : MeasurementListener() {
    private var out: PrintWriter? = null
    private var file: File? = null
    private val FOLDER: String
    private var currentFileName: String? = null
    override fun initResources() {
        val formatter = SimpleDateFormat("yyyy_MM_dd_HH_mm_ss")
        val now = formatter.format(Date())
        val fileName = String.format(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOCUMENTS).toString() + "%s_%s.txt", "$FOLDER/GNSS_Log", now)
        currentFileName = "GNSS_Log_$now.txt"
        file = createFile(fileName)
        try {
            out = PrintWriter(file)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    override fun releaseResources() {
        if (out != null) out!!.close()
        if (SharedData.instance.filesFragment != null) SharedData.instance.filesFragment!!.addFileView(currentFileName)
    }

    override fun write(s: String?) {
        if (out != null && SharedData.instance.isListeningForMeasurements) out!!.println(s)
    }

    private fun createFile(s: String): File {
        val f = File(s)
        if (!f.parentFile.exists()) {
            f.parentFile.mkdirs()
        }
        try {
            f.createNewFile()
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return f
    }

    init {
        FOLDER = SharedData.instance.context!!.getString(R.string.app_documents_folder)
    }
}